src/core/geometry/qgsgeometry.cpp

```diff
@@ -1590,6 +1596,52 @@ bool QgsGeometry::convertToMultiType()
   return true;
 }
 
+bool QgsGeometry::convertToCurvedMultiType()
+{
+  if ( !d->geometry )
+  {
+    return false;
+  }
+
+  switch ( QgsWkbTypes::flatType( d->geometry->wkbType() ) )
+  {
+    case QgsWkbTypes::MultiPoint:
+    case QgsWkbTypes::MultiCurve:
+    case QgsWkbTypes::MultiSurface:
+    case QgsWkbTypes::GeometryCollection:
+    {
+      return true;
+    }
+    default:
+      break;
+  }
+
+  std::unique_ptr< QgsAbstractGeometry >geom = QgsGeometryFactory::geomFromWkbType( QgsWkbTypes::curveType( QgsWkbTypes::multiType( d->geometry->wkbType() ) ) );
+  QgsGeometryCollection *multiGeom = qgsgeometry_cast<QgsGeometryCollection *>( geom.get() );
+  if ( !multiGeom )
+  {
+    return false;
+  }
+
+  QgsGeometryCollection *sourceMultiGeom = qgsgeometry_cast<QgsGeometryCollection *>( d->geometry.get() );
+  if ( sourceMultiGeom )
+  {
+    for ( int i = 0; i < sourceMultiGeom->numGeometries(); ++i )
+    {
+      if ( !multiGeom->addGeometry( sourceMultiGeom->geometryN( i )->clone() ) )
```

Comment:
Potentially we could avoid a bunch of cloning here if we moved the guts of this to the geometry subclasses (e.g. a multilinestring could just transfer the line parts into the compound curve result parts), but we can defer that until this method sees more widespread use.

Reply:
agreed